### PR TITLE
Don't allow changing backdated date with lines assigned.

### LINKED ISSUE #5146: Cannot backdate prescriptions - "stock not available" when it should be
## What went wrong? 😲

Video attached -> [here](https://drive.google.com/drive/folders/1vVZi9r26J4W4NIYqGNyWa30mAjvphiHt). It's long and hefty but captures at least 3 errors?

- Stock of 878 should be available for today and yesterday. And it is! But requires a bit of a workaround, as on the Ledger you will be able to see 877 issued in a prescription. However the system returns errors when trying to issue that quantity.
- When there is an error, it should show up once and not have duplicates occurring at the same time
- Available quantity showed as double (timestamp 2:28)

What concerns me is if you set a smaller value (5 instead of 878), it works fine both ways. It’s only if you set a higher number issues occur both ways. 877 or 828 causes issues too. Only one stock line for the item I was testing (e.g. only one batch). 

Database files attached ->>> [here](https://mega.nz/fm/NEw0lBCT)

## Expected behaviour 🤔

Able to backdate prescriptions without random errors occuring.

## How to Reproduce 🔨

Steps to reproduce the behaviour:

1. Select an item that had some stock movement in the past, create an Outbound Shipment or a Prescription for this item for today.
2. Create a second prescription with that same item -> try to backdate it -> see if you run into same errors.
3. Might be quite hard to recreate so attaching my database above and in the video you can see an item I was testing. 

## Your environment 🌱

- Open mSupply Version: V2.3.0-RC9
- Legacy mSupply Central Server Version: V7.17.10

- Platform: Tablet (apk)

- Database type: SQLite3

### LINKED ISSUE #5153: Prevent changing prescription date, once lines are allocated
## What went wrong? 😲

If we change the date of a prescription, it's hard to correctly account for stock movements as the stocklines have already been modified be the invoice, so if we allow changing of prescription date we need to somehow undo the stockline changes, but also correctly account for the stock movements which might or might not be in the date range for the new invoice date.

## Expected behaviour 🤔

If a user tries to change the invoice date for a prescription after it's had lines added, they will get a prompt

Changing the prescription date will remove all lines in this prescription, do you want to continue?

If they select yes, all lines will be deleted and the date changed.

## How to Reproduce 🔨

Steps to reproduce the behaviour:

1. Create a prescription
2. Add a line
3. Try to change the date
4. See prompt
5. Check Yes, and No both work as expected.

## Your environment 🌱

- Open mSupply Version:
- Legacy mSupply Central Server Version:

- Platform:

- Database type:

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #5153
Fixes #5146

# 👩🏻‍💻 What does this PR do?

Stops users from changing prescription date when lines are assigned.

If they do they'll get a prompt asking if they want to delete lines and continue.

image

## 💌 Any notes for the reviewer?

# 🧪 Testing

- Create a prescription
- Add a line
- Try to change the date
- See prompt
- Check Yes, and No both work as expected.

# 📃 Documentation

- [ ] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [X] **These areas should be updated or checked**: 
  1.Backdate Prescription

### CODE REVIEW

File: server/service/src/invoice/prescription/update/validate.rs [hunk 1]
```diff
@@ -39,9 +37,16 @@ pub fn validate(
...
-        // Check that any lines already assigned won't create a ledger discrepancy
-        check_stock_available_at_date(connection, &invoice.id, backdated_datetime)?;
+    if patch.backdated_datetime.is_some() {
+        // Check if we have any lines allocated to this invoice, if so we can't backdate
+        let lines = InvoiceLineRepository::new(connection).query_by_filter(
+            InvoiceLineFilter::new().invoice_id(EqualFilter::equal_to(&patch.id)),
+        )?;
+        if lines.len() > 0 {
+            return Err(CantBackDate(
+                "Can't Invoice has allocated lines".to_string(),
```
Comment (jmbrunskill) [inline server/service/src/invoice/prescription/update/validate.rs:47]:
```suggestion
                "Can't backdate as invoice has allocated lines".to_string(),
```

File: server/service/src/invoice/prescription/update/validate.rs [hunk 2]
```diff
@@ -39,9 +37,16 @@ pub fn validate(
...
 
-    if let Some(backdated_datetime) = &patch.backdated_datetime {
-        // Check that any lines already assigned won't create a ledger discrepancy
-        check_stock_available_at_date(connection, &invoice.id, backdated_datetime)?;
+    if patch.backdated_datetime.is_some() {
+        // Check if we have any lines allocated to this invoice, if so we can't backdate
+        let lines = InvoiceLineRepository::new(connection).query_by_filter(
+            InvoiceLineFilter::new().invoice_id(EqualFilter::equal_to(&patch.id)),
+        )?;
+        if lines.len() > 0 {
```
Comment (andreievg) [inline server/service/src/invoice/prescription/update/validate.rs:45]: there is count() method for InvoiceLineRepository

Comment (jmbrunskill) [reply server/service/src/invoice/prescription/update/validate.rs:45]: Nice, good point will change...

**Review by andreievg — CHANGES_REQUESTED** [review]
If i click into the date and click out of it i get the delete lines msg, I suspect it should only happen when date is changed ?

**Comment by jmbrunskill** [comment]
> If i click into the date and click out of it i get the delete lines msg, I suspect it should only happen when date is changed ?

On it!

**Comment by jmbrunskill** [comment]
Thanks @lache-melvin for the assist!

**Comment by lache-melvin** [comment]
Hold up, for some reason "onCancel" is getting called on confirm...

File: client/packages/invoices/src/Prescriptions/DetailView/Toolbar.tsx [hunk 3]
```diff
@@ -76,22 +127,16 @@ export const Toolbar: FC = () => {
             />
           </Box>
           <Box display="flex" flexDirection="column" flex={1} marginLeft={3}>
+            <Typography></Typography>
```
Comment (andreievg) [inline client/packages/invoices/src/Prescriptions/DetailView/Toolbar.tsx:130]: Is there a need for empty typography here ?

**Comment by lache-melvin** [comment]
> Hold up, for some reason "onCancel" is getting called on confirm...

Oh man this is our default ConfirmationModal behaviour 😭

**Review by andreievg — APPROVED** [review]
Looks great thanks.

Would have been nice if DateTimePickerInput would only call updates when value changed from default etc..

**Comment by andreievg** [comment]
Maybe next time we go for the bare minimum change ? (an error message from backend and page/state refresh)

**Review by andreievg — APPROVED** [review]
Looks good, @lache-melvin  mentioned she tried another couple of datepickers and they didn't explode

**Comment by lache-melvin** [comment]
> they didn't explode

a low bar 😁

**Pull request merged**